### PR TITLE
BUG: Fix for .str.replace with repl function

### LINKED ISSUE #15055: .str.replace does not accept a repl function
I wanted to pass a function as `repl` to `.str.replace` so it was used by `re.sub`. But it seems that a bug fix for #13438 prevented that to happen. What are your thoughts about this?

#### Code Sample

```python
s = pd.Series(['foo', 'bar', 'baz'])
s.str.replace(r'[aeiou]', lambda m: m.group().upper())
```

#### Problem description

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-26-bc58c5c60bd2> in <module>()
----> 1 s.str.replace(r'[aeiou]', lambda m: m.group().upper())

/anaconda3/lib/python3.4/site-packages/pandas/core/strings.py in replace(self, pat, repl, n, case, flags)
   1504     def replace(self, pat, repl, n=-1, case=True, flags=0):
   1505         result = str_replace(self._data, pat, repl, n=n, case=case,
-> 1506                              flags=flags)
   1507         return self._wrap_result(result)
   1508 

/anaconda3/lib/python3.4/site-packages/pandas/core/strings.py in str_replace(arr, pat, repl, n, case, flags)
    320     # Check whether repl is valid (GH 13438)
    321     if not is_string_like(repl):
--> 322         raise TypeError("repl must be a string")
    323     use_re = not case or len(pat) > 1 or flags
    324 

TypeError: repl must be a string
```

#### Expected Output

    0    fOO
    1    bAr
    2    bAz
    dtype: object

#### Output of ``pd.show_versions()``

<details>

    INSTALLED VERSIONS
    ------------------
    commit: None
    python: 3.4.5.final.0
    python-bits: 64
    OS: Linux
    OS-release: 2.6.32-573.3.1.el6.x86_64
    machine: x86_64
    processor: x86_64
    byteorder: little
    LC_ALL: en_US.UTF-8
    LANG: en_US.UTF-8
    LOCALE: en_US.UTF-8
    
    pandas: 0.19.0
    nose: 1.3.7
    pip: 9.0.1
    setuptools: 18.1
    Cython: 0.25.1
    numpy: 1.11.1
    scipy: 0.17.1
    statsmodels: 0.6.1
    xarray: None
    IPython: 5.0.0
    sphinx: 1.3.1
    patsy: 0.4.0
    dateutil: 2.4.2
    pytz: 2015.7
    blosc: None
    bottleneck: 1.0.0
    tables: 3.2.2
    numexpr: 2.6.1
    matplotlib: 1.5.0
    openpyxl: 2.3.2
    xlrd: 0.9.4
    xlwt: None
    xlsxwriter: None
    lxml: 3.5.0
    bs4: 4.4.1
    html5lib: None
    httplib2: None
    apiclient: None
    sqlalchemy: 1.0.14
    pymysql: None
    psycopg2: 2.6.1 (dt dec pq3 ext)
    jinja2: 2.8
    boto: 2.38.0
    pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by hzpc-joostk**
.str.replace now accepts a callable (function) as replacement string. It now raises a TypeError when repl is not string like nor a callable. Docstring updated accordingly.

  - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry
- [x] closes #15055

### CODE REVIEW

**Comment by chris-b1** [comment]
Looks good, can you add tests and a whatsnew entry?

File: pandas/core/strings.py [hunk 1]
```diff
@@ -303,8 +303,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
     ----------
     pat : string
         Character sequence or regular expression
-    repl : string
-        Replacement sequence
+    repl : string or function
+        Replacement string or a function, which passed the match object and
```
Comment (jreback) [inline pandas/core/strings.py]: function -> callable

**Comment by hzpc-joostk** [comment]
Thanks, @chris-b1 and @jreback. Where should I put the whatsnew entry? Can I just pick an empty spot in doc/source/whatsnew/v0.20.0.txt?

    - Bug in ``.str.replace`` now accepts a callable replacement which is passed to ``re.sub`` (:issue:`15055`)

This maybe a new feature or improvement rather than a bug, right?

**Comment by chris-b1** [comment]
Yeah, I think new features is fine - I had marked the issue a  bug because this would have worked in older versions, although basically undocumented.

File: pandas/tests/test_strings.py [hunk 2]
```diff
@@ -435,6 +435,12 @@ def test_replace(self):
                 for data in (['a', 'b', None], ['a', 'b', 'c', 'ad']):
                     values = klass(data)
                     self.assertRaises(TypeError, values.str.replace, 'a', repl)
+        
+        # GH 15055, callable repl
+        repl = lambda m: m.group(0).swapcase()
+        result = values.str.replace('[a-z][A-Z]{2}', repl, n=2)
+        exp = Series([u('foObaD__baRbaD'), NA])
```
Comment (jreback) [inline pandas/tests/test_strings.py]: test where the callable has different than 1 arg (should be error)

test using named field in re (and callable works correctly)

Comment (hzpc-joostk) [inline pandas/tests/test_strings.py]: Thanks, @jreback . Named field/group is a good idea.

If the callable raises a `TypeError` because of a wrong number of arguments, this is caught in [`pandas.core.strings._map`](https://github.com/hzpc-joostk/pandas/blob/pandas-GH15055-patch-1/pandas/core/strings.py#L175). 

We could overcome this by either inspecting the callable signature: `inspect.signature` or trying with a dummy Regex Match object. But this might have the side effect that the callable is called once before doing the replace on the array, confusing users with some kind of counter within the callable.

Sorry about the checks going wrong. It seems I am too careless with checking my commits. 😕

Comment (hzpc-joostk) [inline pandas/tests/test_strings.py]: With `inspect` I could do

```py
n_empty_args = sum(1 for p in inspect.signature(repl).parameters.values() 
                   if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD, p.KEYWORD_ONLY) 
                   and p.default is p.empty)

# or better readable syntax

signature = inspect.signature(repl)
empty_args = [p for p in signature.parameters.values() 
              if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD, p.KEYWORD_ONLY) 
              and p.default is p.empty]
n_empty_args = len(empty_args)

# followed by

if n_empty_args > 1:
    raise TypeError("callable should have at most one positional without a default")
```

`n_empty_args != 1` might work, unless the first arg has a default for some reason.

Comment (jreback) [inline pandas/tests/test_strings.py]: this is too complicated, instead *maybe* you can introspect the actual error message in ``_map`` and if its a certain class of errors, just re-raise the TypeError

File: pandas/tests/test_strings.py [hunk 3]
```diff
@@ -435,6 +435,31 @@ def test_replace(self):
                 for data in (['a', 'b', None], ['a', 'b', 'c', 'ad']):
                     values = klass(data)
                     self.assertRaises(TypeError, values.str.replace, 'a', repl)
+        
+        ## GH 15055
```
Comment (jreback) [inline pandas/tests/test_strings.py]: can you make a separate test

File: pandas/core/strings.py [hunk 4]
```diff
@@ -167,7 +167,14 @@ def _map(f, arr, na_mask=False, na_value=np.nan, dtype=object):
         try:
             convert = not all(mask)
             result = lib.map_infer_mask(arr, f, mask.view(np.uint8), convert)
-        except (TypeError, AttributeError):
+        except (TypeError, AttributeError) as e:
+            re_missing = (r'missing \d+ required (positional|keyword-only) '
```
Comment (jreback) [inline pandas/core/strings.py]: can you add a 1-line comment on what types of exceptions you are re-raising here (imagine you are a future reader and are like what?)

File: doc/source/whatsnew/v0.20.0.txt [hunk 5]
```diff
@@ -23,6 +23,7 @@ New features
 ~~~~~~~~~~~~
 
 - Integration with the ``feather-format``, including a new top-level ``pd.read_feather()`` and ``DataFrame.to_feather()`` method, see :ref:`here <io.feather>`.
+- ``.str.replace`` now accepts a callable replacement which is passed to ``re.sub`` (:issue:`15055`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: callable, as a replacement, which ....

File: pandas/core/strings.py [hunk 6]
```diff
@@ -318,9 +326,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
     """
 
     # Check whether repl is valid (GH 13438)
-    if not is_string_like(repl):
-        raise TypeError("repl must be a string")
-    use_re = not case or len(pat) > 1 or flags
+    if not (is_string_like(repl) or callable(repl)):
+        raise TypeError("repl must be a string or function")
```
Comment (jreback) [inline pandas/core/strings.py]: function -> callable

File: pandas/core/strings.py [hunk 7]
```diff
@@ -303,8 +310,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
     ----------
     pat : string
         Character sequence or regular expression
-    repl : string
-        Replacement sequence
+    repl : string or callable
+        Replacement string or a callable, it's passed the match object and 
+        must return a replacement string to be used. See :func:`re.sub`.
```
Comment (jreback) [inline pandas/core/strings.py]: add the version (0.20.0) after callable

File: pandas/core/strings.py [hunk 8]
```diff
@@ -303,8 +310,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
...
     n : int, default -1 (all)
         Number of replacements to make from start
```
Comment (jreback) [inline pandas/core/strings.py]: can you add some examples to the doc-string?

also would be nice to have an example in text.rst

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. just some added documenations / minor comments. ping on green.

**Comment by hzpc-joostk** [comment]
Thanks. I added changes based on your review. Only building with python=2.7 fails. I have little to no experience with Python 2x. I didn't caught the TypeError in Python 2:

**Comment by jreback** [comment]
so

- some linting issues, use ``git diff master | flake8 --diff`` to show locally
- setup a venv or conda env with py2.7 to step thru the issues there.

Generally python 3 does the right thing with exceptions, while py2 may not be re-raising exceptions correctly (so your catching may not work here).

So you may need to have a slightly different strategy in py2 and py3 for catching the invalid args.

If you get stuck, ping.

**Comment by hzpc-joostk** [comment]
Currently I'm designing new regexes for both py27 and py3k. I'll look into the linting.

**Comment by jreback** [comment]
@hzpc-joostk 

>Currently I'm designing new regexes for both py27 and py3k. I'll look into the linting.

what do you mean? you can prob do something very straightforward, you don't even need regexes, you are just trying to find out if the error message if related to the calling of the function (and not the dtype of the data).

**Comment by hzpc-joostk** [comment]
@jreback earlier you reviewed a part of my code:
> test where the callable has different than 1 arg (should be error)

In `.core.strings._map` these string methods are mapped. On both `TypeError` and `AttributeError`, `NaN` is returned instead. Using regex, implemented in commit 27065a29ba883f79499e4978872611f23d606f7e, only `TypeError`s regarding wrong number of args are reraised. However, these regexes only worked for Python 3. With e15dcdf544f02de04573b942132186d8ea397676, these regexes are unified into one sophisticated (complex) regex and work for Python 2 (tested).

tl;dr the tests should pass, as this now raises an error.

**Comment by jreback** [comment]
@hzpc-joostk lgtm. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by hzpc-joostk** [comment]
Now fixing linting issues: 

    C:\Users\joostk\Documents\git\pandas>git diff 6ecc43d83801bd0337b771390ea5f4063689f275 | flake8 --diff
    .\pandas/core/strings.py:313:75: W291 trailing whitespace
    .\pandas/core/strings.py:314:70: W291 trailing whitespace
    .\pandas/core/strings.py:332:65: W291 trailing whitespace
    .\pandas/core/strings.py:341:69: W291 trailing whitespace
    .\pandas/core/strings.py:342:71: W291 trailing whitespace
    .\pandas/tests/test_strings.py:438:1: W293 blank line contains whitespace
    .\pandas/tests/test_strings.py:439:1: D102 Missing docstring in public method
    .\pandas/tests/test_strings.py:440:9: E266 too many leading '#' for block comment
    .\pandas/tests/test_strings.py:468:1: D102 Missing docstring in public method

**Comment by hzpc-joostk** [comment]
@jreback Thanks for your input and help. All changes made. Waiting for CI... Besides that: all green!

**Comment by jreback** [comment]
https://travis-ci.org/pandas-dev/pandas/jobs/194493716

@hzpc-joostk 

so you can use

```
if PY2:
    # match py2 excetion
else:
    # match PY3 exception
```

and as I said you can do a minimal matching, because sometimes these exception messages change in versions (e.g. 3.5 -> 3.6 for example, maybe not in this case, but the wording can change). You are just generically interested in a specific type of error message (and not its specific wording per se)

**Comment by jreback** [comment]
thanks @hzpc-joostk 

nicely done!

**Comment by hzpc-joostk** [comment]
Thanks @jreback 

Awesome. My first (accepted) PR on GitHub. 🍾 Cheers
